File: vscode-nexus/src/repl.ts
/**
 * Interactive Nexus REPL (tree-walk or bytecode VM).
 */

import * as readline from "readline";
import * as path from "path";
import { evaluate } from "./compiler/engine";
import { findStdlibDir, MODULES_DIRNAME } from "./runtime/modules";
import { BytecodeCompiler } from "./vm/compiler";
import { VirtualMachine } from "./vm/vm";
import { BUILTIN_NAMES } from "./language/builtins";
import { SymbolTable } from "./vm/symbols";
import { NexusObject, NULL_OBJ } from "./language/values";
import { Lexer } from "./language/lexer";
import { Parser } from "./language/parser";
import { lowerSyntax } from "./language/syntax";
import { MacroExpander } from "./language/macro";

export interface ReplConfig {
  engine?: "eval" | "vm";
  rootDir?: string;
  stdlibDir?: string;
  prompt?: string;
  input?: NodeJS.ReadableStream;
  output?: NodeJS.WritableStream;
}

export async function runRepl(cfg: ReplConfig = {}): Promise<void> {
  const engine = cfg.engine ?? "eval";
  const rootDir = cfg.rootDir ?? process.cwd();
  const stdlibDir =
    cfg.stdlibDir !== undefined ? cfg.stdlibDir : findStdlibDir(rootDir);
  const prompt = cfg.prompt ?? "nex> ";
  const input = cfg.input ?? process.stdin;
  const output = cfg.output ?? process.stdout;

  output.write("Nexus REPL — :help for commands, :quit to exit\n");
  output.write(`engine=${engine}\n`);

  let pending = "";
  let currentEngine = engine;

  // VM persistent state
  let constants: NexusObject[] = [];
  let globals: (NexusObject | undefined)[] = new Array(65536);
  let symbolTable = new SymbolTable();
  for (let i = 0; i < BUILTIN_NAMES.length; i++) {
    symbolTable.defineBuiltin(i, BUILTIN_NAMES[i]!);
  }

  const rl = readline.createInterface({ input, output, terminal: true });

  const ask = (): Promise<string | null> =>
    new Promise((resolve) => {
      rl.question(pending ? "...  " : prompt, (line) => {
        resolve(line);
      });
      rl.once("close", () => resolve(null));
    });

  while (true) {
    const line = await ask();
    if (line === null) {
      output.write("\n");
      break;
    }
    const trim = line.trim();

    if (!pending && trim.startsWith(":")) {
      if (trim === ":quit" || trim === ":exit") {
        break;
      }
      if (trim === ":help") {
        output.write(
          "Commands: :help :quit :engine eval|vm :clear\n",
        );
        continue;
      }
      if (trim.startsWith(":engine")) {
        const parts = trim.split(/\s+/);
        const next = parts[1];
        if (next === "eval" || next === "vm") {
          currentEngine = next;
          output.write(`engine=${currentEngine}\n`);
        } else {
          output.write("usage: :engine eval|vm\n");
        }
        continue;
      }
      if (trim === ":clear") {
        pending = "";
        constants = [];
        globals = new Array(65536);
        symbolTable = new SymbolTable();
        for (let i = 0; i < BUILTIN_NAMES.length; i++) {
          symbolTable.defineBuiltin(i, BUILTIN_NAMES[i]!);
        }
        output.write("state cleared\n");
        continue;
      }
      output.write(`unknown command ${trim}\n`);
      continue;
    }

    pending = pending ? `${pending}\n${line}` : line;
    if (!isBalanced(pending)) {
      continue;
    }
    const source = pending;
    pending = "";

    try {
      if (currentEngine === "vm") {
        const lowered = lowerSyntax(source);
        const lexer = new Lexer(lowered);
        const parser = new Parser(lexer);
        let program = parser.parseProgram();
        if (parser.getErrors().length > 0) {
          output.write(parser.getErrors().join("\n") + "\n");
          continue;
        }
        program = new MacroExpander().expand(program).program;
        const compiler = new BytecodeCompiler(symbolTable, constants);
        const err = compiler.compile(program);
        if (err) {
          output.write(err + "\n");
          continue;
        }
        const bytecode = compiler.bytecode();
        constants = bytecode.constants;
        symbolTable = compiler.symbolTable;
        const vm = new VirtualMachine(bytecode, globals);
        const runErr = vm.run();
        globals = vm.globals;
        for (const o of vm.output) {
          output.write(o + "\n");
        }
        if (runErr) {
          output.write(`ERROR: ${runErr}\n`);
        } else {
          const v = vm.lastPoppedStackElem();
          if (v !== NULL_OBJ && v.type !== "NULL") {
            output.write(v.inspect() + "\n");
          }
        }
      } else {
        const result = await evaluate(source, {
          tier: "eval",
          rootDir,
          modulesDir: path.join(rootDir, MODULES_DIRNAME),
          stdlibDir,
          checkOwnership: false,
          enableEffects: false,
        });
        for (const o of result.output) {
          output.write(o + "\n");
        }
        if (result.value.type === "ERROR") {
          output.write(result.value.inspect() + "\n");
        } else if (result.value.type !== "NULL") {
          output.write(result.value.inspect() + "\n");
        }
      }
    } catch (err) {
      output.write(
        `ERROR: ${err instanceof Error ? err.message : String(err)}\n`,
      );
    }
  }

  rl.close();
}

function isBalanced(source: string): boolean {
  let braces = 0;
  let parens = 0;
  let brackets = 0;
  let inString = false;
  for (let i = 0; i < source.length; i++) {
    const ch = source[i]!;
    if (ch === '"' && source[i - 1] !== "\\") {
      inString = !inString;
      continue;
    }
    if (inString) {
      continue;
    }
    if (ch === "{") {
      braces++;
    } else if (ch === "}") {
      braces--;
    } else if (ch === "(") {
      parens++;
    } else if (ch === ")") {
      parens--;
    } else if (ch === "[") {
      brackets++;
    } else if (ch === "]") {
      brackets--;
    }
  }
  return braces <= 0 && parens <= 0 && brackets <= 0;
}

File: pkg/repl/repl.go
// Package repl implements an interactive Nexus read-eval-print loop.
package repl

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"

	"nex-lang/pkg/compiler"
	"nex-lang/pkg/evaluator"
	"nex-lang/pkg/lexer"
	"nex-lang/pkg/parser"
	"nex-lang/pkg/runtime"
	"nex-lang/pkg/vm"
)

// Config controls REPL behavior.
type Config struct {
	// Engine is "tree" (default) or "vm".
	Engine    string
	RootDir   string
	StdlibDir string
	Prompt    string
}

// Run starts the REPL on in/out until EOF or :quit.
func Run(in io.Reader, out io.Writer, cfg Config) error {
	if cfg.Prompt == "" {
		cfg.Prompt = "nex> "
	}
	if cfg.Engine == "" {
		cfg.Engine = "tree"
	}
	if cfg.RootDir == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return err
		}
		cfg.RootDir = cwd
	}

	fmt.Fprintln(out, "Nexus REPL — :help for commands, :quit to exit")
	fmt.Fprintf(out, "engine=%s\n", cfg.Engine)

	scanner := bufio.NewScanner(in)
	var pending strings.Builder

	env := evaluator.NewEnvironment()
	rt := runtime.NewWithOptions(cfg.RootDir, env, runtime.Options{StdlibDir: cfg.StdlibDir})

	// VM/REPL state
	constants := []evaluator.Object{}
	globals := make([]evaluator.Object, vm.GlobalsSize)
	symbolTable := compiler.NewSymbolTable()
	for i, name := range evaluator.BuiltinNames {
		symbolTable.DefineBuiltin(i, name)
	}

	for {
		if pending.Len() == 0 {
			fmt.Fprint(out, cfg.Prompt)
		} else {
			fmt.Fprint(out, "...  ")
		}
		if !scanner.Scan() {
			fmt.Fprintln(out)
			break
		}
		line := scanner.Text()
		trim := strings.TrimSpace(line)

		if pending.Len() == 0 && strings.HasPrefix(trim, ":") {
			if done, err := handleCommand(trim, out, &cfg); done {
				return err
			}
			continue
		}

		pending.WriteString(line)
		pending.WriteByte('\n')
		src := pending.String()
		if !isComplete(src) {
			continue
		}
		pending.Reset()

		switch cfg.Engine {
		case "vm":
			if err := evalVM(src, out, &constants, globals, &symbolTable); err != nil {
				fmt.Fprintf(out, "error: %v\n", err)
			}
		default:
			result := rt.LoadSource(src, "<repl>")
			if errObj, ok := result.(*evaluator.Error); ok {
				fmt.Fprintf(out, "error: %s\n", errObj.Message)
				continue
			}
			if result != nil && result != evaluator.NULL {
				fmt.Fprintln(out, result.Inspect())
			}
		}
	}
	return scanner.Err()
}

func handleCommand(cmd string, out io.Writer, cfg *Config) (quit bool, err error) {
	parts := strings.Fields(cmd)
	switch parts[0] {
	case ":quit", ":exit", ":q":
		return true, nil
	case ":help", ":h":
		fmt.Fprint(out, `Commands:
  :help           Show this help
  :quit           Exit the REPL
  :engine tree|vm Switch evaluation engine
  :clear          Clear pending multi-line input (also Ctrl+C in most terminals)

Multi-line input is accepted until braces/brackets/parens are balanced.
`)
	case ":engine":
		if len(parts) < 2 || (parts[1] != "tree" && parts[1] != "vm") {
			fmt.Fprintln(out, "usage: :engine tree|vm")
			return false, nil
		}
		cfg.Engine = parts[1]
		fmt.Fprintf(out, "engine=%s\n", cfg.Engine)
	case ":clear":
		fmt.Fprintln(out, "ok")
	default:
		fmt.Fprintf(out, "unknown command %q — try :help\n", parts[0])
	}
	return false, nil
}

func evalVM(src string, out io.Writer, constants *[]evaluator.Object, globals []evaluator.Object, st **compiler.SymbolTable) error {
	l := lexer.New(src)
	p := parser.New(l)
	program := p.ParseProgram()
	if len(p.Errors()) > 0 {
		return fmt.Errorf("%s", strings.Join(p.Errors(), "; "))
	}
	comp := compiler.NewWithState(*st, *constants)
	if err := comp.Compile(program); err != nil {
		return err
	}
	*constants = comp.Constants()
	*st = comp.SymbolTable()
	machine := vm.NewWithGlobalsState(comp.Bytecode(), globals)
	if err := machine.Run(); err != nil {
		return err
	}
	result := machine.LastPoppedStackElem()
	if result != nil && result != evaluator.NULL {
		fmt.Fprintln(out, result.Inspect())
	}
	return nil
}

func isComplete(src string) bool {
	var paren, brace, bracket int
	inStr := false
	escape := false
	for _, r := range src {
		if inStr {
			if escape {
				escape = false
				continue
			}
			if r == '\\' {
				escape = true
				continue
			}
			if r == '"' {
				inStr = false
			}
			continue
		}
		switch r {
		case '"':
			inStr = true
		case '(':
			paren++
		case ')':
			paren--
		case '{':
			brace++
		case '}':
			brace--
		case '[':
			bracket++
		case ']':
			bracket--
		}
	}
	return paren <= 0 && brace <= 0 && bracket <= 0 && !inStr
}
